### PR TITLE
TestStorageMongoDB using strikememongo

### PR DESCRIPTION
**Pull request by altergui**
a simple way of addressing #16

### CODE REVIEW

**Comment by altergui** [comment]
the test is failing with "E11000 duplicate key error dup key" because ms.NewAttempt() is trying to `_, err = ms.tokenIndex.InsertOne(ctx, atindex)`
i'm looking into this

**Comment by altergui** [comment]
this error is not related to strikememongo, a live mondodb server v5.0.9 also gives the same result (with a slightly more verbose error)

` [E11000 duplicate key error collection: xhextvwhckovbua.tokenindex index: _id_ dup key: { _id: BinData(0, E8D0A7CAAEFB43688E090209138AD439) }]" `

File: handlers/smshandler/mongodbstorage.go [hunk 1]
```diff
@@ -261,7 +261,9 @@ func (ms *MongoStorage) NewAttempt(userID, electionID types.HexBytes,
 	}
 	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
 	defer cancel()
-	_, err = ms.tokenIndex.InsertOne(ctx, atindex)
+	if r := ms.tokenIndex.FindOne(ctx, bson.M{"_id": token}); r.Err() != nil {
```
Comment (jordipainan) [inline handlers/smshandler/mongodbstorage.go:264]: I would advocate for specifically check that Err is `ErrNoDocuments` as FindOne can return other types of errors and can be useful to act differently depending on the type

File: handlers/smshandler/mongodbstorage.go [hunk 2]
```diff
@@ -261,7 +261,9 @@ func (ms *MongoStorage) NewAttempt(userID, electionID types.HexBytes,
 	}
 	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
 	defer cancel()
-	_, err = ms.tokenIndex.InsertOne(ctx, atindex)
```
[the remark this answers is not in the document]
Comment (p4u) [reply handlers/smshandler/mongodbstorage.go:264]: If the token is repeated it means the test is wrong and should be adapted.

File: handlers/smshandler/mongodbstorage.go [hunk 3]
```diff
@@ -261,7 +261,9 @@ func (ms *MongoStorage) NewAttempt(userID, electionID types.HexBytes,
...
+	if r := ms.tokenIndex.FindOne(ctx, bson.M{"_id": token}); r.Err() != nil {
+		_, err = ms.tokenIndex.InsertOne(ctx, atindex)
```
Comment (jordipainan) [inline handlers/smshandler/mongodbstorage.go:265]: Seems that this error is ignored

**Review by jordipainan — COMMENTED** [review]
Added some minor suggestions :)

**Comment by altergui** [comment]
ok thanks for your comments, i now understand the logic behind it and also understood why the JSON storage handler didn't fail the same test: the Set() function in that case "Set adds a key-value pair. If the key already exists, its value is updated."

https://github.com/vocdoni/blind-csp/blob/b84d231f6105a13317fb9ddc811f610e78d9d91b/handlers/smshandler/jsonstorage.go#L192-L195

so i fixed the testStorage() as indicated

**Comment by p4u** [comment]
thanks for improving the test :heart:

**Pull request merged**